conda/gateways/repodata/__init__.py

```diff
@@ -289,3 +317,434 @@ def conda_http_errors(url, repodata_fn):
+        # read repodata.json
+        # check stat, if wrong clear cache information
+
+        with self.cache_path_state.open("r+") as state_file, _lock(state_file):
+            # cannot use pathlib.read_text / write_text on any locked file, as
+            # it will release the lock early
+            state = json.loads(state_file.read())
+
+            # json and state files should match. must read json before checking
+            # stat (if json_data is to be trusted)
+            if state_only:
+                json_data = ""
+            else:
+                json_data = self.cache_path_json.read_text()
+
+            json_stat = self.cache_path_json.stat()
+            if not (
+                state.get("mtime_ns") == json_stat.st_mtime_ns
+                and state.get("size") == json_stat.st_size
+            ):
+                # clear mod, etag, cache_control to encourage re-download
+                state.update({"etag": "", "mod": "", "cache_control": "", "size": 0})
+            self.state.clear()
+            self.state.update(
+                state
+            )  # will aliased _mod, _etag (not cleared above) pass through as mod, etag?
+
+        return json_data
+
+        # check repodata.json stat(); mtime_ns must equal .state.json, or it is stale
+        # read repodata.json
+        # check repodata.json stat() again: st_size, st_mtime_ns must be equal
+
+        # repodata.json is okay - use it somewhere
+
+        # repodata.json is not okay - maybe use it, but don't allow cache updates
+
+        # unlock .state.json
+
+        # also, add refresh_ns instead of touching repodata.json file
+
+    def load_state(self):
+        """
+        Update self.state without reading repodata.json.
+
+        Return self.state.
+        """
+        try:
+            self.load(state_only=True)
+        except FileNotFoundError:
+            self.state.clear()
+        return self.state
+
+    def save(self, data: str):
+        """
+        Write data to <repodata>.json cache path, synchronize state.
+        """
+        temp_path = self.cache_dir / self.cache_dir.with_suffix(f".{os.urandom(4).hex()}.tmp")
+
+        try:
+            with temp_path.open("x") as temp:  # exclusive mode, error if exists
+                temp.write(data)
+
+            return self.replace(temp_path)
+
+        finally:
+            try:
+                temp_path.unlink()
+            except OSError:
+                pass
+
+    def replace(self, temp_path: Path):
+        """
+        Rename path onto <repodata>.json path, synchronize state.
+
+        Relies on path's mtime not changing on move. `temp_path` should be
+        adjacent to `self.cache_path_json` to be on the same filesystem.
+        """
+        with self.cache_path_state.open("a+") as state_file, _lock(state_file):
+            # "a+" avoids trunctating file before we have the lock and creates
+            state_file.seek(0)
+            state_file.truncate()
```

Comment:
Since how the file is opened is important should we perhaps be doing the `open` within the lock context? something like:

```suggestion
        with _lock(self.cache_path_state) as state_file:
```

Reply:
Maybe? I feel like there may have been a reason not to.

Reply:
The difference between `r+` and `a+` that matters to us, is that `r+` errors if the file does not exist while `a+` will create the file, and seek to the end of the file. `w+` will truncate and seek to the beginning, which is not something we want to do before we have the lock.